File: examples/java/pact_validation_with_oas/one_consumer_one_provider/consumer/src/test/java/br/com/zup/pact/consumer/service/impl/ClientServiceImplTest.java
package br.com.zup.pact.consumer.service.impl;

import br.com.zup.pact.consumer.dto.BalanceDTO;
import br.com.zup.pact.consumer.dto.ClientDetailsDTO;
import br.com.zup.pact.consumer.integration.account.service.AccountIntegrationService;
import br.com.zup.pact.consumer.repository.ClientRepository;
import br.com.zup.pact.consumer.service.ClientService;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;

import java.math.BigDecimal;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.Mockito.when;

@SpringBootTest
public class ClientServiceImplTest {

    @MockBean
    private ClientRepository clientRepository;

    @MockBean
    private AccountIntegrationService accountIntegrationService;

    @Autowired
    private ClientService clientService;

    @Test
    void getClientDetailsTest() {

        when(this.clientRepository.findByClientId(anyInt()))
                .thenReturn(Optional.of(this.clientDetailsDTOStub()));

        assertEquals(this.clientService.getClientDetails(1), Optional.of(this.clientDetailsDTOStub()));
    }

    @Test
    void getAllClientsDetailsListTest() {

        when(this.clientRepository.getAll())
                .thenReturn(Optional.of(this.clientDetailsDTOStubList()));

        assertEquals(this.clientService.getAll(), Optional.of(this.clientDetailsDTOStubList()));
    }

    @Test
    void getBalanceTest() {
        final BalanceDTO balanceDTO = BalanceDTO.builder()
                .accountId(1)
                .clientId(1)
                .balance(new BigDecimal("100.0"))
                .build();

        when(this.clientRepository.findByClientId(anyInt()))
                .thenReturn(Optional.of(this.clientDetailsDTOStub()));

        when(this.accountIntegrationService.getBalance(anyInt()))
                .thenReturn(Optional.of(balanceDTO));

        assertEquals(this.clientService.getBalance(1), Optional.of(balanceDTO));
    }

    private ClientDetailsDTO clientDetailsDTOStub() {
        return ClientDetailsDTO.builder()
                .id(1)
                .accountId(1)
                .name("Any First Name")
                .finalName("Any Final Name")
                .age(20)
                .build();
    }

    private List<ClientDetailsDTO> clientDetailsDTOStubList() {
        return Arrays.asList(
                ClientDetailsDTO.builder()
                        .id(1)
                        .accountId(1)
                        .name("Any First Name")
                        .finalName("Any Final Name")
                        .age(20)
                        .build(),
                ClientDetailsDTO.builder()
                        .id(2)
                        .accountId(2)
                        .name("Another First Name")
                        .finalName("Another Final Name")
                        .age(26)
                        .build()
        );
    }
}

File: example/java/kotlin/consumer/src/test/kotlin/br/com/zup/pact/consumer/service/ClientServiceImplTest.kt
package br.com.zup.pact.consumer.service

import br.com.zup.pact.consumer.dto.BalanceDTO
import br.com.zup.pact.consumer.dto.ClientDetailsDTO
import br.com.zup.pact.consumer.exception.ClientNotFoundException
import br.com.zup.pact.consumer.integration.account.service.AccountIntegrationService
import br.com.zup.pact.consumer.repository.ClientRepository
import io.mockk.clearMocks
import io.mockk.every
import io.mockk.junit5.MockKExtension
import io.mockk.mockk
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import java.util.*

@ExtendWith(MockKExtension::class)
class ClientServiceImplTest {

    private val clientRepositoryMock: ClientRepository = mockk()
    private val accountIntegrationServiceMock: AccountIntegrationService = mockk()
    private val clientService: ClientService = ClientServiceImpl(
            clientRepositoryMock,
            accountIntegrationServiceMock
    )

    @BeforeEach
    fun setUp() {
        clearMocks(clientRepositoryMock)
    }

    @Test
    fun `Method getClientDetails should return findByClientId from the repository`() {

        every { clientRepositoryMock.findByClientId(1) }
                .returns(Optional.of(createClientDetailsDTO(accountId = 1)))

        val expectedAccountDetailsDTO = clientService.getClientDetails(1)

        Assertions.assertThat(expectedAccountDetailsDTO).isEqualTo(
                Optional.of(createClientDetailsDTO(accountId = 1))
        )
    }

    @Test
    fun `Method getAll should return getAll from the repository`() {

        every { clientRepositoryMock.getAll() }
                .returns(listOf(createClientDetailsDTO(accountId = 1)))

        val expectedAccountDetailsDTO = clientService.getAll()

        Assertions.assertThat(expectedAccountDetailsDTO).isEqualTo(
                listOf(createClientDetailsDTO(accountId = 1))
        )
    }

    @Test
    fun `Method getBalance should return Optional of BalanceDTO`() {

        every { clientRepositoryMock.findByClientId(1) }
                .returns(Optional.of(createClientDetailsDTO(accountId = 2)))

        every { accountIntegrationServiceMock.getBalance(2) }
                .returns(Optional.of(BalanceDTO(
                        accountId = 2,
                        balance = 100.0
                )))

        val expectedBalanceDTO = clientService.getBalance(1)

        Assertions.assertThat(expectedBalanceDTO).isEqualTo(
                Optional.of(BalanceDTO(
                        accountId = 2,
                        balance = 100.0
                ))
        )
    }

    @Test
    fun `Method getBalance should throw ClientNotFoundException when ClientDetails is null`() {

        every { clientRepositoryMock.findByClientId(1) }
                .returns(Optional.empty())

        Assertions.assertThatExceptionOfType(ClientNotFoundException::class.java)
                .isThrownBy { clientService.getBalance(1) }
    }

    private fun createClientDetailsDTO(
            accountId: Int,
            id: Int = 100,
            name: String = "First name",
            finalName: String = "Final name",
            age: Int = 18
    ) = ClientDetailsDTO(
            id,
            accountId,
            name,
            finalName,
            age
    )
}